### PR TITLE
Add NUnit capability

### LINKED ISSUE #418: Ability to convert NUnit results into BHoM Framework
Following workshop with @alelom :

Create NUnit_oM and NUnit_Engine to support deserialisation of NUnit test results into the BHoM framework and eco-system.

### PR DESCRIPTION
**Pull request by FraserGreenroyd**
### Issues addressed by this PR
Fixes #418 

 ### Test files

 ### Changelog

 ### Additional comments

### CODE REVIEW

**Comment by FraserGreenroyd** [comment]
@BHoMBot check compliance

**Comment by FraserGreenroyd** [comment]
@BHoMBot check compliance

**Comment by FraserGreenroyd** [comment]
@BHoMBot check compliance

**Comment by FraserGreenroyd** [comment]
@BHoMBot fix copyright headers ref. 11329381789

**Comment by FraserGreenroyd** [comment]
@BHoMBot check copyright-compliance
@BHoMBot check project-compliance

**Comment by FraserGreenroyd** [comment]
@BHoMBot fix project file ref. 11329433267

**Comment by FraserGreenroyd** [comment]
@BHoMBot fix copyright headers ref. 11329428713

**Comment by FraserGreenroyd** [comment]
@BHoMBot check copyright-compliance

File: NUnit_Engine/Compute/RunTests.cs [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+            MemoryStream stm = new MemoryStream();
+
+            StreamWriter stw = new StreamWriter(stm);
+            stw.Write(testResult.OuterXml);
+            stw.Flush();
+
+            stm.Position = 0;
+
+            XmlSerializer ser = new XmlSerializer(typeof(TestRun));
+            return (ser.Deserialize(stm) as TestRun);
```
Comment (alelom) [inline NUnit_Engine/Compute/RunTests.cs:38]: XmlDocument xmlDoc = new XmlDocument();
	xmlDoc.LoadXml(testResult.OuterXml);

	XmlSerializer ser = new XmlSerializer(typeof(TestRun));
	using (XmlNodeReader reader = new XmlNodeReader(xmlDoc))
	{
		result = ser.Deserialize(reader) as TestRun;
	}
```suggestion
            MemoryStream stm = new MemoryStream();

            StreamWriter stw = new StreamWriter(stm);
            stw.Write(testResult.OuterXml);
            stw.Flush();

            stm.Position = 0;

            XmlSerializer ser = new XmlSerializer(typeof(TestRun));
            return (ser.Deserialize(stm) as TestRun);
```

File: NUnit_oM/Assertion.cs [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+using System.Text;
+using System.Threading.Tasks;
+using System.Xml.Serialization;
+
+namespace BH.oM.Test.NUnit
+{
+    [XmlRoot(ElementName = "assertion")]
+    public class Assertion
+    {
+
```
Comment (alelom) [inline NUnit_oM/Assertion.cs:35]:
```suggestion
```

**Comment by FraserGreenroyd** [comment]
@BHoMBot check compliance

**Comment by FraserGreenroyd** [comment]
@BHoMBot fix copyright headers ref. 11358412516

**Comment by FraserGreenroyd** [comment]
@BHoMBot check copyright-compliance

**Comment by FraserGreenroyd** [comment]
@BHoMBot fix copyright headers ref. 11358447388

**Comment by FraserGreenroyd** [comment]
@BHoMBot check copyright-compliance

**Comment by FraserGreenroyd** [comment]
@BHoMBot check compliance

**Comment by FraserGreenroyd** [comment]
@BHoMBot fix project file ref. 11360908814

**Comment by FraserGreenroyd** [comment]
@BHoMBot check compliance

**Comment by FraserGreenroyd** [comment]
@BHoMBot check null-handling

File: NUnit_Engine/Convert/ResultStatus.cs [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+        [Input("resultStatus", "The NUnit result string, typically either 'Passed' or 'Failed'.")]
+        [Output("testStatus", "The BHoM TestStatus enum value converted from the NUnit result string.")]
+        public static TestStatus ToTestStatus(string resultStatus)
+        {
+            if (resultStatus == "Failed")
+                return TestStatus.Error;
+            else if (resultStatus == "Passed")
+                return TestStatus.Pass;
+
+            return TestStatus.Error; //Default - something probably went wrong so we'd want to investigate
```
Comment (alelom) [inline NUnit_Engine/Convert/ResultStatus.cs:45]: Just wondering if we have a "Warning" state

File: NUnit_oM/Environment.cs [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+        public virtual string Culture { get; set; }
+
+        [XmlAttribute(AttributeName = "uiculture")]
+        public virtual string Uiculture { get; set; }
+
+        [XmlAttribute(AttributeName = "os-architecture")]
+        public virtual string OsArchitecture { get; set; }
+    }
+}
+
```
Comment (alelom) [inline NUnit_oM/Environment.cs:69]: I'm surprised by your lack of decency
```suggestion
```

Comment (FraserGreenroyd) [reply NUnit_oM/Environment.cs:69]: 🤣

Comment (alelom) [reply NUnit_oM/Environment.cs:69]: You just dismissed all my previous suggested changes you naughty - trailing newlines in all files! I wonder where we will end up

File: NUnit_Engine/Convert/ResultStatus.cs (hunk 3 above)
Comment (FraserGreenroyd) [reply NUnit_Engine/Convert/ResultStatus.cs:45]: We do have a `Warning` option on the `TestStatus` however, they get swallowed up and hidden by the bot - so I prefer to stick with `Error` to highlight we might need to investigate something. Like an early warning system.

File: NUnit_Engine/Compute/RunTests.cs [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+            TestRun result = null;
+            using (XmlNodeReader reader = new XmlNodeReader(xmlDoc))
+            {
+                result = ser.Deserialize(reader) as TestRun;
+            }
+            return result;
+        }
+    }
+}
+
```
Comment (FraserGreenroyd) [inline NUnit_Engine/Compute/RunTests.cs:65]:
```suggestion
```

File: NUnit_Engine/Convert/ResultStatus.cs [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+        }
+    }
+}
+
```
Comment (FraserGreenroyd) [inline NUnit_Engine/Convert/ResultStatus.cs:49]:
```suggestion
```

File: NUnit_Engine/Convert/TestResult.cs [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+            result.Message = $"";
+
+            if(nunitTestCase.Failure != null)
+                result.Message += $"{nunitTestCase.Failure.Message}{System.Environment.NewLine}{nunitTestCase.Failure.Stacktrace}";
+
+            return result;
+        }
+    }
+}
+
```
Comment (FraserGreenroyd) [inline NUnit_Engine/Convert/TestResult.cs:92]:
```suggestion
```

File: NUnit_oM/Assertion.cs [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+        public virtual string Stacktrace { get; set; }
+
+        [XmlAttribute(AttributeName = "result")]
+        public virtual string Result { get; set; }
+
+        [XmlText]
+        public virtual string Text { get; set; }
+    }
+}
+
```
Comment (FraserGreenroyd) [inline NUnit_oM/Assertion.cs:48]:
```suggestion
```

File: NUnit_oM/Assertions.cs [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+namespace BH.oM.Test.NUnit
+{
+    [XmlRoot(ElementName = "assertions")]
+    public class Assertions
+    {
+        [XmlElement(ElementName = "assertion")]
+        public virtual Assertion Assertion { get; set; }
+    }
+}
+
```
Comment (FraserGreenroyd) [inline NUnit_oM/Assertions.cs:39]:
```suggestion
```

File: NUnit_oM/Failure.cs [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+    public class Failure
+    {
+        [XmlElement(ElementName = "message")]
+        public virtual string Message { get; set; }
+
+        [XmlElement(ElementName = "stack-trace")]
+        public virtual string Stacktrace { get; set; }
+    }
+}
+
```
Comment (FraserGreenroyd) [inline NUnit_oM/Failure.cs:42]:
```suggestion
```

File: NUnit_oM/Properties.cs [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+namespace BH.oM.Test.NUnit
+{
+    [XmlRoot(ElementName = "properties")]
+    public class Properties
+    {
+        [XmlElement(ElementName = "property")]
+        public virtual List<Property> Property { get; set; }
+    }
+}
+
```
Comment (FraserGreenroyd) [inline NUnit_oM/Properties.cs:39]:
```suggestion
```

File: NUnit_oM/Property.cs [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+    public class Property
+    {
+        [XmlAttribute(AttributeName = "name")]
+        public virtual string Name { get; set; }
+
+        [XmlAttribute(AttributeName = "value")]
+        public virtual string Value { get; set; }
+    }
+}
+
```
Comment (FraserGreenroyd) [inline NUnit_oM/Property.cs:42]:
```suggestion
```

File: NUnit_oM/TestSuite.cs [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+        public virtual Environment Environment { get; set; }
+
+        [XmlElement(ElementName = "settings")]
+        public virtual Settings Settings { get; set; }
+
+        [XmlElement(ElementName = "properties")]
+        public virtual Properties Properties { get; set; }
+    }
+}
+
```
Comment (FraserGreenroyd) [inline NUnit_oM/TestSuite.cs:115]:
```suggestion
```

**Comment by FraserGreenroyd** [comment]
@BHoMBot check compliance
@BHoMBot check null-handling

**Review by alelom — APPROVED** [review]
Approving based on code review, and the development session done with @FraserGreenroyd.

**Pull request merged**